pkg/controller/dmcluster/dm_cluster_control.go

```diff
@@ -0,0 +1,197 @@
+	conditionUpdater DMClusterConditionUpdater
+	recorder         record.EventRecorder
+}
+
+// UpdateStatefulSet executes the core logic loop for a dmcluster.
+func (dcc *defaultDMClusterControl) UpdateDMCluster(dc *v1alpha1.DMCluster) error {
+	dcc.defaulting(dc)
+	if !dcc.validate(dc) {
+		return nil // fatal error, no need to retry on invalid object
+	}
+
+	var errs []error
+	oldStatus := dc.Status.DeepCopy()
+
+	if err := dcc.updateDMCluster(dc); err != nil {
+		errs = append(errs, err)
+	}
+
+	if err := dcc.conditionUpdater.Update(dc); err != nil {
+		errs = append(errs, err)
+	}
+
+	if apiequality.Semantic.DeepEqual(&dc.Status, oldStatus) {
+		return errorutils.NewAggregate(errs)
+	}
+	if _, err := dcc.dcControl.UpdateDMCluster(dc.DeepCopy(), &dc.Status, oldStatus); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errorutils.NewAggregate(errs)
+}
+
+func (dcc *defaultDMClusterControl) defaulting(dc *v1alpha1.DMCluster) {
+	defaulting.SetDMClusterDefault(dc)
+}
+
+func (dcc *defaultDMClusterControl) validate(dc *v1alpha1.DMCluster) bool {
+	errs := v1alpha1validation.ValidateDMCluster(dc)
+	if len(errs) > 0 {
+		aggregatedErr := errs.ToAggregate()
+		klog.Errorf("dm cluster %s/%s is not valid and must be fixed first, aggregated error: %v", dc.GetNamespace(), dc.GetName(), aggregatedErr)
+		dcc.recorder.Event(dc, v1.EventTypeWarning, "FailedValidation", aggregatedErr.Error())
+		return false
+	}
+	return true
+}
+
+func (dcc *defaultDMClusterControl) updateDMCluster(dc *v1alpha1.DMCluster) error {
+	// TODO: implement reclaimPolicyManager
+	// syncing all PVs managed by operator's reclaim policy to Retain
+	// if err := dcc.reclaimPolicyManager.Sync(dc); err != nil {
+	//	return err
+	// }
+
+	// TODO: add orphanPodsCleaner for dm cluster
+	// cleaning all orphan pods(pd, tikv or tiflash which don't have a related PVC) managed by operator
+	// skipReasons, err := dcc.orphanPodsCleaner.Clean(dc)
+	// if err != nil {
+	// 	return err
+	// }
+	// if klog.V(10) {
+	// 	for podName, reason := range skipReasons {
+	// 		klog.Infof("pod %s of cluster %s/%s is skipped, reason %q", podName, dc.Namespace, dc.Name, reason)
+	// 	}
+	// }
+
+	// TODO: restarted pods in dm cluster
+	// sync all the pods which need to be restarted
+	// if err := dcc.podRestarter.Sync(dc); err != nil {
+	// 	return err
+	// }
+
+	// works that should do to making the dm-master cluster current state match the desired state:
+	//   - create or update the dm-master service
+	//   - create or update the dm-master headless service
+	//   - create the dm-master statefulset
+	//   - sync dm-master cluster status from dm-master to DMCluster object
+	//   - set two annotations to the first dm-master member:
+	// 	   - label.Bootstrapping
+	// 	   - label.Replicas
+	//   - upgrade the dm-master cluster
+	//   - scale out/in the dm-master cluster
+	//   - failover the dm-master cluster
+	if err := dcc.masterMemberManager.Sync(dc); err != nil {
```

Comment:
Can we just log and append the error here and continue to sync the worker? WDYT @weekface @cofyc